### PR TITLE
HDDS-11290. Container scanner should keep scanning after non-fatal errors

### PR DESCRIPTION
**Pull request by errose28**
### Motivation

In order to build the merkle tree of all data in the container, the scanner should not exit after the first issue it encounters like it does currently. The scanner should track and return all errors that it sees, and only stop the scan on fatal errors that prevent further scanning of the container, like DB access errors.

This PR is a pre-requisite to HDDS-10374. It does not actually generate a merkle tree during the scan and is also not testing this functionality. It sets up HDDS-10374 to be an easy drop in to the scanner which will allow the focus of that change to be testing of merkle tree generation.

### Primary Changes

Previously `ScanResult` was an object that encapsulated a singe error. This was the first error the scanner saw which would abort the scan. This change decouples the `ScanResult` from the errors, which are now represented by a list of`ContainerScanError`s in the `ScanResult`.
- `ScanResult` is a general interface that can represent a data or metadata scan. It can be logged by entities like the `ContainerLogger` which do not care where the unhealthy result came from.
- `MetadataScanResult` is a `ScanResult` implementation produced from a container metadata scan.
  - This scan will not produce a merkle tree since it does not check data.
- `DataScanResult` extends `MetadataScanResult` by adding a merkle tree representing the data that was scanned.
  - All data scans begin with a metadata scan, and then proceed to scan the data only if the metadata scan succeeds.
- The datanode application log will only log the first error  encountered unless debug logging is enabled, then all errors will be logged.
  - The container log does not have this option to keep it from rolling off.

### Secondary Changes

- General cleanup of `KeyValueContainerCheck` internals were done since this PR already required invasive changes in this area.
- Fixed a bug from #5485/HDDS-9005 where the DB was being used to check for container deletion during a scan instead of the container state in memory.
  - This would fail to detect a schema v2 container which is completely lost but not actually deleted. The container would remain in the datanode's memory as healthy.
  - Since the [memory state is updated first](https://github.com/apache/ozone/blob/769a9aad1fe8c936c5dc5d5019ca4ed61644c042/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java#L1512) and our container object will still be valid even after being removed from the [ContainerSet](https://github.com/apache/ozone/blob/769a9aad1fe8c936c5dc5d5019ca4ed61644c042/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/OzoneContainer.java#L114), we can use the `DELETED` state as the deletion check which is both safer and simpler.
- When a container is deleted during a scan, this is indicated with a different field in the `ScanResult`. It is no longer considered an error and will not cause the `ScanResult` to be unhealthy.
- Container log checks are more stringent now making sure that the error logged is actually for the expected container ID.
  - Since the tests re-use the same cluster, one could have passed if a previous test logged the failure type this test was expecting.

### Notes to Reviewers

It is probably best to review the scan flow end-to-end instead in addition to just viewing the diff.

## What is the link to the Apache JIRA

HDDS-11290

## How was this patch tested?

The scanner has a lot of existing tests that should all pass ensuring no regressions:
- `TestKeyValueContainerCheck`
- `TestKeyValueHandlerWithUnhealthyContainer`
- `Test{Background,OnDemand}Container{Data,Metadata}Scanner`
- `Test{Background,OnDemand}Container{Data,Metadata}ScannerIntegration`

New tests added for detecting multiple errors will be added in HDDS-11471 since this already a large PR.

### CODE REVIEW

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java [hunk 1]
```diff
@@ -87,27 +87,30 @@ public void scanContainer(Container<?> c)
...
-          containerId, result.getException());
-      return;
-    }
-    if (!result.isHealthy()) {
-      LOG.error("Corruption detected in container [{}]. Marking it UNHEALTHY.",
-          containerId, result.getException());
-      boolean containerMarkedUnhealthy = controller.markContainerUnhealthy(containerId, result);
-      if (containerMarkedUnhealthy) {
-        metrics.incNumUnHealthyContainers();
+    DataScanResult result = c.scanData(throttler, canceler);
```
Comment (kerneltime) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java:90]: Nit: Not part of this PR but the whole `scanContainer` should have a metric around time taken

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java [hunk 2]
```diff
@@ -87,27 +87,30 @@ public void scanContainer(Container<?> c)
...
+
+    if (result.isDeleted()) {
+      LOG.debug("Container [{}] has been deleted during the data scan.", containerId);
+    } else {
+      if (!result.isHealthy()) {
+        logUnhealthyScanResult(containerId, result, LOG);
+
+        // Only increment the number of unhealthy containers if the container was not already unhealthy.
+        if (controller.markContainerUnhealthy(containerId, result)) {
+          metrics.incNumUnHealthyContainers();
```
Comment (kerneltime) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java:100]: This method does not check if before the scan the container was healthy, then before incrementing the metric we should check if there is a state transition or if the container was already unhealthy.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java (hunk 1 above)
Comment (kerneltime) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java:90]: NIT:
```
    if (volume.isFailed()) {
      shutdown("The volume has failed.");
      return;
    }
``` 
We should log volume info in the log message in line 80.

**Review by kerneltime — COMMENTED** [review]
One pass, will give it another pass along with tests next.

**Comment by errose28** [comment]
[CI is green on my fork](https://github.com/errose28/ozone/tree/HDDS-11290-scan-on-error). I'll rebase and take this out of draft. New unit tests of multiple failure scenarios will be done in HDDS-11471 due to the size of this PR already.

**Comment by kerneltime** [comment]
Let me know once conflicts are resolved.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java [hunk 3]
```diff
@@ -78,7 +78,7 @@ public abstract class ContainerData {
   private String chunksPath;
 
   // State of the Container
-  private ContainerDataProto.State state;
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java:81]: Synchronized lets multiple threads call those methods safely. That part is already in the code. Volatile makes those threads read the latest value when they call the methods. It is necessary for correctness.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java (hunk 1 above)
Comment (errose28) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java:90]: The full message with the volume info is constructed and logged at INFO level by the `shutdown` method. Is there something extra you think we should add to that message?

Comment (errose28) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java:90]: For the metrics, there are numerous improvements. I filed HDDS-11593 for this.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java (hunk 2 above)
Comment (errose28) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java:100]: I think we should log unhealthy results on each iteration, even if the container was unhealthy on a previous iteration. The reason it is unhealthy may have changed.

For the metric, the `controller.markContainerUnhealthy` call returns a boolean indicating if the container state was changed, which we use to increment the metric. Technically you could have an execution like this:
1. Background scanner marks container unhealthy
2. On-demand scanner moves it back to healthy
3. Background scanner increments the metric

I think this is expected behavior though, because the counts of each container in the metric are from the background scanner's point of view. Since it is a slow running process, states of containers it already scanned can change after it has scanned them but before it has scanned them again.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/OzoneContainer.java [hunk 4]
```diff
@@ -126,6 +126,8 @@ public class OzoneContainer {
   private final ReplicationServer replicationServer;
   private DatanodeDetails datanodeDetails;
   private StateContext context;
+  private final ContainerChecksumTreeManager checksumManager;
```
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/OzoneContainer.java:129]: Good catch, that would have been a nasty bug.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (errose28) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/impl/ContainerData.java:81]: Turns out the thread's cache will be flushed at the exit of the synchronized block. Since all usages are synchronized  we don't need volatile.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java (hunk 2 above)
Comment (errose28) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java:103]: I've added a TODO here about how #7335 will change the way this metric is used when we merge it in from master.

**Review by kerneltime — APPROVED** [review]
